src/lakefs_spec/transaction.py

```diff
@@ -29,112 +27,88 @@
-    """A generic placeholder for a value computed by the lakeFS server in a versioning operation during a transaction."""
-
-    def __init__(self, wrapped: T | None = None):
-        super().__init__(wrapped)
-
-    @property
-    def available(self) -> bool:
-        """Whether the wrapped value is available, i.e. already computed."""
-        return self.__wrapped__ is not None
-
-    @property
-    def value(self):
-        if self.__wrapped__ is None:
-            raise RuntimeError("placeholder unfilled")
-        return self.__wrapped__
-
-    @value.setter
-    def value(self, val: T) -> None:
-        """Fill in the placeholder. Not meant to be called directly except in the completion of the transaction."""
-        self.__wrapped__ = val
+def ensurebranch(b: str | Branch, repository: str, client: Client) -> Branch:
+    if isinstance(b, str):
+        return Branch(repository, b, client=client)
+    return b
 
 
 class LakeFSTransaction(Transaction):
     """
     A lakeFS transaction model capable of versioning operations in between file uploads.
 
+    Creates an ephemeral branch, conducts all uploads and operations on that branch,
+    and optionally merges it back into the source branch on success.
+
     Parameters
     ----------
     fs: LakeFSFileSystem
         The lakeFS file system associated with the transaction.
+    repository: str | Repository
+        The repository in which to conduct the transaction.
+    base_branch: str | Branch
+        The branch on which the resulting files should end up.
+    automerge: bool
+        Automatically merge the ephemeral branch into the base branch after successful
+        transaction completion.
+    delete: bool
+        Delete the ephemeral branch after the transaction.
     """
 
-    def __init__(self, fs: "LakeFSFileSystem"):
+    def __init__(
+        self,
+        fs: "LakeFSFileSystem",
+        repository: str | Repository,
+        base_branch: str | Branch = "main",
+        automerge: bool = True,
+        delete: bool = True,
+    ):
         super().__init__(fs=fs)
         self.fs: "LakeFSFileSystem"
-        self.files: deque[AbstractBufferedFile | VersioningOpTuple] = deque(self.files)
+        self.files: deque[ObjectWriter] = deque(self.files)
+
+        if isinstance(repository, str):
+            self.repository = repository
+        else:
+            self.repository = repository.id
+
+        # base branch needs to be a lakefs.Branch, since it is being diffed
+        # with the ephemeral branch in __exit__.
+        self.base_branch = ensurebranch(base_branch, self.repository, self.fs.client)
+
+        self.automerge = automerge
+        self.delete = delete
+
+        ephem_name = "transaction-" + "".join(random.choices(string.digits, k=6))  # nosec: B311
+        self._ephemeral_branch = Branch(self.repository, ephem_name, client=self.fs.client)
 
     def __enter__(self):
+        self._ephemeral_branch.create(self.base_branch, exist_ok=False)
```

Comment:
How does exception handling look for context managers that fail upon entering (since this has a non-zero probability of failing)? Do you just go straight into `__exit__` with the appropriate exception passed?

Reply:
No, seems like you never make it anywhere, neither the body nor the `__exit__`. Which is actually good in this example, since it does not flip `fs._intrans`, and everything else does not happen (not even cleanup, which is fine, since nothing was created in the first place).

Reply:
Okay, makes sense, in this case we should be good.